### PR TITLE
KAFKA-7502: Cleanup KTable materialization logic in a single place

### PR DESCRIPTION
**Pull request by dongjinleekr**
This is a draft cleanup for KAFKA-7502. Here is the details:

1. Make `KTableKTableJoinNode` abstract, and define its child classes (`[NonMaterialized,Materialized]KTableKTableJoinNode`) instead: now, all materialization-related routines are separated into the other classes.
2. `KTableKTableJoinNodeBuilder#build` now instantiates `[NonMaterialized,Materialized]KTableKTableJoinNode` classes instead of `KTableKTableJoinNode`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by dongjinleekr** [comment]
@guozhangwang Is this what you meant? I found similar routines in `KTableImpl#[filter, map, transform]` methods. If it is correct, I will do the same work to that methods.

Please have a look you are free.

**Comment by dongjinleekr** [comment]
retest this please

**Comment by dongjinleekr** [comment]
Rebased against the latest trunk. cc/ @guozhangwang @mjsax

**Review by guozhangwang — COMMENTED** [review]
@dongjinleekr Thanks for your PR! I went through it and realized my original description on the JIRA ticket may not be very clear. I've hence updated the description and let me know if it clarifies a few things.

As for the PR itself, splitting `KTableKTableJoinNode` seems to me not addressing the original issue, but I do really like the way you inlined `buildJoin`. So how about reducing the scope of this PR to just do the cleanup of inlining `buildJoin`, while we can continue discussing on the ticket at the best solution to achieve the address of the ticket.

**Comment by dongjinleekr** [comment]
I see. Then, how about this? I rebased against the latest trunk and added two commits (296ddfc, e57ad22); Now, all materialization logic in `KTableKTableJoinNode` is now moved into `KTableImpl#doJoin`. Is this way correct?

**Comment by dongjinleekr** [comment]
@guozhangwang I see. Then, how about this? I rebased against the latest trunk and added two commits (296ddfc, e57ad22); Now, all materialization logic in `KTableKTableJoinNode` is now moved into `KTableImpl#doJoin`. Is this way correct?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java [hunk 1]
```diff
@@ -495,57 +477,69 @@ public String queryableStoreName() {
...
+                .withJoinOtherStoreNames(((KTableImpl) other).valueGetterSupplier().storeNames());
 
-        // only materialize if specified in Materialized
         if (materializedInternal != null) {
-            kTableJoinNodeBuilder.withMaterializedInternal(materializedInternal);
+            final Serde<K> keySerde = materializedInternal.keySerde() != null ? materializedInternal.keySerde() : this.keySerde;
+            final Serde<VR> valueSerde = materializedInternal.valueSerde();
+            final String queryableStoreName = materializedInternal.storeName();
+            final StoreBuilder<KeyValueStore<K, VR>> storeBuilder = new KeyValueStoreMaterializer<>(materializedInternal).materialize();
+            kTableJoinNodeBuilder
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: This part and the line 525-529 below can be extracted out of if condition.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java [hunk 2]
```diff
@@ -56,9 +64,39 @@
...
+    public String[] joinThisStoreNames() {
+        return joinThisStoreNames;
+    }
+
+    public String[] joinOtherStoreNames() {
+        return joinOtherStoreNames;
+    }
+
+    public String queryableStoreName() {
+        return queryableStoreName;
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java]: nit: we do not need to save another reference of queryableStoreName at this class, instead we can add a getter for the field with the same name in `KTableKTableJoinMerger` public and just return `joinMergeProcessorParameters.queryableStoreName()` here.

**Review by guozhangwang — COMMENTED** [review]
This PR lgtm. I think this is a good first step towards KAFKA-7502. Left some minor comments.

Call for another review @bbejeck

**Comment by guozhangwang** [comment]
@dongjinleekr could you run the unit test locally and see if the error is related?

**Comment by guozhangwang** [comment]
retest this please

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java [hunk 3]
```diff
@@ -85,44 +123,42 @@ public void writeToTopology(final InternalTopologyBuilder topologyBuilder) {
...
-               ", materializedInternal=" + materializedInternal +
-               "} " + super.toString();
+            "joinThisStoreNames=" + Arrays.toString(joinThisStoreNames()) +
+            ", joinOtherStoreNames=" + Arrays.toString(joinOtherStoreNames()) +
+            ", queryableStoreName=" + queryableStoreName +
+            "} " + super.toString();
     }
 
-    public static <K, V1, V2, VR> KTableKTableJoinNodeBuilder<K, V1, V2, VR> kTableKTableJoinNodeBuilder() {
+    public static <K, V1, V2, VR> KTableKTableJoinNodeBuilder<K, V1, V2, VR> kTableKTableJoinNodeBuilder(final InternalStreamsBuilder builder) {
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java]: The `InternalStreamsBuilder` isn't used and can be removed here. The `InternalStreamsBuilder` gets passed in with the `writeToTopology` method.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java [hunk 4]
```diff
@@ -131,59 +167,71 @@ private KTableKTableJoinNodeBuilder() {
...
-                                              valueJoiner,
-                                              joinThisProcessorParameters,
-                                              joinOtherProcessorParameters,
-                                              joinMergeProcessorParameters,
-                                              materializedInternal,
-                                              thisJoinSide,
-                                              otherJoinSide,
-                                              joinThisStoreNames,
-                                              joinOtherStoreNames);
+                null,
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java]: Why is `valueJoiner` `null` here? Is this intentional?

EDIT: I get it now, the `ValueJoiner` is used in `KTable#doJoin` on lines 483-491 and thus was actually never used by the `KTableKTableJoinNode`.  

Maybe we should alter the constructor of `KTableKTableJoinNode` to not take a `ValueJoiner` at all and explicitly pass  a` null` the parent class on line 60 of `KTableKTableJoinNode`?  

WDYT?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java [hunk 5]
```diff
@@ -33,20 +34,27 @@
...
 public class KTableKTableJoinNode<K, V1, V2, VR> extends BaseJoinProcessorNode<K, Change<V1>, Change<V2>, Change<VR>> {
 
+    private final Serde<K> keySerde;
+    private final Serde<VR> valueSerde;
     private final String[] joinThisStoreNames;
     private final String[] joinOtherStoreNames;
-    private final MaterializedInternal<K, VR, KeyValueStore<Bytes, byte[]>> materializedInternal;
+    // The name of queryableStore, which stores the join results.
+    private final String queryableStoreName;
+    private final StoreBuilder<KeyValueStore<K, VR>> storeBuilder;
```
Comment (bbejeck) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java:39]: Should this get inlined with the types to `StoreBuilder<KeyValueStore<Bytes, byte[]>>`?
\cc @guozhangwang

**Review by bbejeck — COMMENTED** [review]
Overall this looks good to me, just a couple of minor comments.

**Comment by bbejeck** [comment]
Java 8 passed, Java 11 failed but build results already cleaned up.

retest this please

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java (hunk 5 above)
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java:39]: Hmm why can we do that? The storeBuilder returned from the `KeyValueStoreMaterializer` is a fully-wrapped one and not an inner bytes-store, right?

Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/graph/KTableKTableJoinNode.java:39]: Ah, yes that's correct, @dongjinleekr ignore my comment here.

**Comment by bbejeck** [comment]
Both Java 8 and Java 11 failed, test results already cleaned out

retest this please

**Comment by bbejeck** [comment]
@dongjinleekr what's the status on this PR?  Once the comments are addressed I believe we can merge this PR.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 1 above)
Comment (dongjinleekr) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: I tried to, but could not; all parameter variables to `KTableKTableJoinNodeBuilder` must be `final`, but Java does not support conditional instantiation like Scala. 

One alternative approach would be introducing a new class like `MaterializeConfig` to contain them. What do you think?

**Comment by dongjinleekr** [comment]
@bbejeck Here is the update for the `join` operation. I am now working on the other operations and can finish it by this Friday.

**Comment by guozhangwang** [comment]
@dongjinleekr Let's use separate PRs on other operations so that we can keep this one small and tangible for reviewing / merging.

BTW the failures are due to compilation errors: `Execution failed for task ':streams:compileJava'.`

Could you run `./gradlew streams:test` locally and see if you can resolve these errors?

**Comment by dongjinleekr** [comment]
@guozhangwang Here is the fix. I omitted a file from the last commit. (+1. rebased onto the latest trunk.)

And also, I will separate the other operators from this issue in Jira. Is that okay?

cc/ @bbejeck

**Comment by dongjinleekr** [comment]
@guozhangwang Or, only separated PRs for this issue? @bbejeck

**Comment by bbejeck** [comment]
Java 8 failed with 
```
org.apache.kafka.common.security.authenticator.SaslAuthenticatorTest.testValidSaslScramSha256
org.apache.kafka.common.security.authenticator.SaslAuthenticatorTest.testMultipleServerMechanisms
kafka.api.UserQuotaTest.testThrottledProducerConsumer
```
Java 11 passed

retest this please

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java (hunk 1 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KTableImpl.java]: @dongjinleekr you could do something like:
```java
final Serde<K> keySerde;
final Serde<VR> valueSerde;
final String queryableStoreName;
final StoreBuilder<KeyValueStore<K, VR>> storeBuilder;

if (materializedInternal != null){
   ....
else {
   ....
}

kTableJoinNodeBuilder
       .withKeySerde(keySerde)
       .withValueSerde(valueSerde)
       .withQueryableStoreName(queryableStoreName)
       .withStoreBuilder(storeBuilder);
```
Alternatively you can make them non-`final` and set the default values and update the variables inside the `if` condition, and simply remove the `else` block.  It's a matter of personal style so I'll leave it up to you.

**Review by bbejeck — COMMENTED** [review]
This PR looks good to me, less just one minor comment.

**Comment by bbejeck** [comment]
>Or, only separated PRs for this issue?

Fine with me to have multiple PRs on one Jira ticket, just make sure to have the Jira number in each PR title and in the PR comments state what parts you are working on.
\cc @guozhangwang

**Comment by dongjinleekr** [comment]
Here it is. OMG, I just learned one more thing about Java's final variable! Thanks @bbejeck :)

And about Issue management, I am +1 to separating the PRs only. In that case, two of the other PRs should be based on the other, since `TableProcessorNode` is used by all of `filter`, `map`, and `transform`. (In other words, one former PR will be the preliminary work for the other two PRs.)

I am now working on `filter` now. Let's roll! :smile:

cc/ @guozhangwang

**Review by bbejeck — APPROVED** [review]
LGTM! Thanks @dongjinleekr

**Comment by guozhangwang** [comment]
LGTM!

**Pull request merged**